gmqtt/message_queue.py

```diff
@@ -0,0 +1,125 @@
+
+    def match(self, topic: str) -> bool:
+        topic_levels = topic.split("/")
+
+        for filter_level, topic_level in zip(self.levels, topic_levels):
+            if filter_level == "+":
+                continue
+            elif filter_level == "#":
+                return True
+            else:
+                if filter_level != topic_level:
+                    return False
+        return True
+
+    def __hash__(self):
+        return hash("/".join(self.levels))
+
+
+class Subscription:
+    def __init__(self, message_queue: asyncio.Queue, on_unsubscribe: Callable):
+        self._incoming_messages = message_queue
+        self._on_unsubscribe = on_unsubscribe
+
+    async def recv(self):
+        """Receive the next message published to this subscription"""
+        message = await self._incoming_messages.get()
+        # TODO: Hold off sending PUBACK for `message` until this point
+        return message
+
+    def __aiter__(self):
+        return self
+
+    async def __anext__(self):
+        return await self.recv()
+
+    async def unsubscribe(self):
+        await self._on_unsubscribe()
+
+    def _add_message(self, message: Message):
+        self._incoming_messages.put_nowait(message)
+
+
+class DropPolicy(enum.Enum):
+    OLDEST_FIRST = enum.auto()
+
+
+class SubscriptionManager:
+    """
+    Manages incoming messages and the downstream subscription objects.
+
+    Handles:
+      - Maximum Queue size and message drop policy
+      - Routing messages to subscriptions based on topic (possibily with wildcards)
+    """
+
+    def __init__(
+        self, receive_maximum: int, drop_policy: DropPolicy = DropPolicy.OLDEST_FIRST
+    ):
+        self.receive_maximum = receive_maximum
+        self.drop_policy = drop_policy
+        self.subs: Dict[TopicFilter, List["asyncio.Queue"]] = collections.defaultdict(
+            list
+        )
+        self.size = 0
+
+    async def add_subscription(self, topic_filter_str: str) -> Subscription:
+        topic_filter = TopicFilter(topic_filter_str)
+
+        subscribed_messages = asyncio.Queue(self.receive_maximum)
+        sub_id = (topic_filter, len(self.subs[topic_filter]))
+        self.subs[topic_filter].append(subscribed_messages)
+
+        return Subscription(
+            message_queue=subscribed_messages,
+            on_unsubscribe=functools.partial(self.remove_subscription, sub_id=sub_id),
+        )
+
+    async def remove_subscription(self, sub_id: Tuple[TopicFilter, int]):
+        topic_filter, idx = sub_id
+
+        # TODO: Unsubscribe on underlying client
+        # Note: Don't remove so indexes are preserved
+        self.subs[topic_filter][idx] = None
+
+    def on_message(self, message: Message):
+        # TODO: check self.size
+
+        # if over, attempt to drop qos=0 packet using `drop_policy`
+
+        # if under, add to appropiate queues:
+        for subscription_topic in self.subs:
```

Comment:
It's all good, but we can reduce count of indents and make the code more beautiful.

```python
for topic in filter(lambda x: x.match(message.topic)):
    for sub in self.subs[topic]:
        subs.put_nowait(message)
```
or 

```python
for subscription_topic in self.subs:
    if not subscription_topic.match(message.topic):
        continue
    for subscription in self.subs[subscription_topic]:
        subscription.put_nowait(message)
```